File: 6.binary-search/동규/3.search-in-rotated-sorted-array.js
// https://leetcode.com/problems/search-in-rotated-sorted-array/

var search = function (nums, target) {
  let left = 0,
    right = nums.length - 1;
  while (left < right) {
    const mid = left + Math.floor((right - left) / 2);
    if (nums[mid] === target) return mid;
    // When middle element is less than the last element
    if (nums[mid] < nums[right]) {
      if (target > nums[mid] && target <= nums[right]) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }
    // When middle element is greater than the last element
    else {
      if (target > nums[mid] || target < nums[left]) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }
  }
  return left;
};

File: 6.binary-search/연경/3.search-in-rotated-sorted-array.py
## https://leetcode.com/problems/search-in-rotated-sorted-array/
from typing import List
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # You must write an algorithm with O(log n) runtime complexity.
        l, r = 0, len(nums) - 1
        while l <= r:
            m = (l + r) // 2
            if nums[m] == target:
                return m

            if nums[l] <= nums[m]:
                if nums[l] <= target <= nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                if nums[m] <= target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1

        return -1

if __name__ == "__main__":
    nums = [4,5,6,7,0,1,2]
    target = 3
    print(Solution().search(nums, target))